File: 101-symmetric-tree.cpp
// https://leetcode.com/problems/symmetric-tree

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool check(TreeNode* l, TreeNode* r){
        if(!l && !r)
            return true;
        if(!l && r)
            return false;
        if(l && !r)
            return false;
        if(l->val != r->val)
            return false;
        
        return check(l->left, r->right) && check(l->right, r->left);
    }
    bool isSymmetric(TreeNode* root) {
        if(!root)
            return true;
        return check(root->left, root->right);
    }
};

File: 101-symmetric-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def is_mirror(self, l: Optional[TreeNode], r: Optional[TreeNode]):
        if not l and not r:
            return True
        if not l or not r or l.val != r.val:
            return False
        return self.is_mirror(l.left, r.right) and self.is_mirror(l.right, r.left)

    def isSymmetric(self, root: Optional[TreeNode]) -> bool:
        return not root or self.is_mirror(root.left, root.right)
